File: android/app/src/main/java/com/synergyboat/flashcardAi/core/benchmark/ExecutionLogger.kt
package com.synergyboat.flashcardAi.core.benchmark

import java.util.logging.Logger
import javax.inject.Inject
import javax.inject.Singleton
import kotlin.system.measureTimeMillis

@Singleton
class ExecutionLogger @Inject constructor(
    private val logger: Logger
) {
    suspend fun <T> logExecDuration(
        action: suspend () -> T,
        name: String = "no_name",
        tag: String = "no_tag",
        log: Boolean = true
    ): T {
        var result: T
        val elapsedTime = measureTimeMillis {
            result = action()
        }

        if (log) {
            logger.info("$tag | Execution time for $name: $elapsedTime ms")
        }

        return result
    }
}

File: ios/FlashcardAI/Core/Benchmark/ExecutionLogger.swift
import Foundation

final class ExecutionLogger {
    @discardableResult
    static func logExecDuration<T>(name: String = "no_name", tag: String = "no_tag", log: Bool = true, _ action: () throws -> T) rethrows -> T {
        let start = Date()
        let result = try action()
        let elapsed = Date().timeIntervalSince(start) * 1000.0
        if log {
            print("\(tag) | Execution time for \(name): \(String(format: "%.0f", elapsed)) ms")
        }
        return result
    }

    static func logExecDurationAsync<T>(name: String = "no_name", tag: String = "no_tag", log: Bool = true, _ action: @escaping () async throws -> T) async rethrows -> T {
        let start = Date()
        let result = try await action()
        let elapsed = Date().timeIntervalSince(start) * 1000.0
        if log {
            print("\(tag) | Execution time for \(name): \(String(format: "%.0f", elapsed)) ms")
        }
        return result
    }
}


